File: iOS/DateMate/UI/MyProfile/MyProfileViewModel.swift
//
//  MyProfileViewModel.swift
//  DateMate
//
//  Created by 홍희표 on 2025/12/01.
//

import Foundation
import Combine

class MyProfileViewModel: ObservableObject {
    private let userRepository: UserRepository
    
    private let postRepository: PostRepository
    
    @Published var uiState = MyProfileUiState()
    
    private var cancellables = Set<AnyCancellable>()
    
    private func loadUserProfile() {
        userRepository.remoteUserStatePublisher
            .compactMap { $0 }
            .sink { user in
                self.loadProfile(userId: user.uid)
                self.loadPosts(userId: user.uid)
            }
            .store(in: &cancellables)
    }
    
    private func loadProfile(userId: String) {
        userRepository.fetchUserProfile(userId: userId)
            .sink { resource in
                switch resource.state {
                case .Loading:
                    self.uiState.isLoading = true
                case .Success:
                    self.uiState.profile = resource.data!
                    self.uiState.isLoading = false
                case .Error:
                    self.uiState.message = resource.message
                    self.uiState.isLoading = false
                }
            }
            .store(in: &cancellables)
    }
    
    private func loadPosts(userId: String) {
        postRepository.fetchUserPosts(userId: userId)
            .sink { resource in
                switch resource.state {
                case .Loading:
                    self.uiState.isLoading = true
                case .Success:
                    let posts = resource.data ?? []
                    self.uiState.posts = posts
                    self.uiState.stats = UserStats(
                        postCount: posts.count,
                        followers: "2.5M",
                        following: "259"
                    )
                    self.uiState.isLoading = false
                case .Error:
                    self.uiState.message = resource.message
                    self.uiState.isLoading = false
                }
            }
            .store(in: &cancellables)
    }
    
    init(_ userRepository: UserRepository, _ postRepository: PostRepository) {
        self.userRepository = userRepository
        self.postRepository = postRepository
        
        loadUserProfile()
    }
}

File: Android/app/src/main/java/com/hhp227/datemate/ui/myprofile/MyProfileViewModel.kt
package com.hhp227.datemate.ui.myprofile

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.datemate.common.Resource
import com.hhp227.datemate.data.repository.PostRepository
import com.hhp227.datemate.data.repository.ProfileRepository
import com.hhp227.datemate.data.repository.UserRepository
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class MyProfileViewModel(
    private val userRepository: UserRepository,
    private val profileRepository: ProfileRepository,
    private val postRepository: PostRepository
) : ViewModel() {
    private val _uiState = MutableStateFlow<MyProfileUiState>(MyProfileUiState())
    val uiState: StateFlow<MyProfileUiState> = _uiState.asStateFlow()

    private fun loadUserProfile() {
        viewModelScope.launch {
            userRepository.remoteUserStateFlow
                .filterNotNull()
                .flatMapLatest { user ->
                    val profileFlow = profileRepository.fetchUserProfile(user.uid)
                    val postsFlow = postRepository.fetchUserPosts(user.uid)

                    combine(profileFlow, postsFlow) { profileResource, postsResource ->
                        if (profileResource is Resource.Loading || postsResource is Resource.Loading) {
                            return@combine MyProfileUiState(isLoading = true)
                        }
                        if (profileResource is Resource.Success && postsResource is Resource.Success) {
                            val profile = profileResource.data
                            val posts = postsResource.data ?: emptyList()
                            val stats = UserStats(
                                postCount = posts.size,
                                followers = "2.5M", // 추후 팔로우 기능 구현 시 교체
                                following = "259"
                            )
                            return@combine MyProfileUiState(profile = profile, stats = stats, posts = posts)
                        } else {
                            return@combine MyProfileUiState(message = "프로필 정보를 불러올 수 없습니다.")
                        }
                    }
                }
                .collect { state ->
                    _uiState.value = state
                }
        }
    }

    init {
        loadUserProfile()
    }
}